docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -20,196 +45,193 @@ The scenarios of defining placement rules in SQL include:
-ALTER TABLE table_name
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=1,
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1;
+CREATE TABLE t1 (
+	id INT NOT NULL PRIMARY KEY,
+	b VARCHAR(100)
+) PLACEMENT CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=1, CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1, CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1;
 ```
 
-This statement indicates PD to schedule the leader to `sh`, add one follower to  `sh` and one to `gz`. Note that as the leader can be only one, the first clause doesn't actually add a replica, so this statement adds 2 replicas.
-
-`ADD PLACEMENT POLICY` also supports adding TiFlash replicas for a table, as statement `ALTER TABLE table_name SET TIFLASH REPLICA count` does. For example:
+Adding `PLACEMENT` to an existing table:
 
 ```sql
-ALTER TABLE table_name
-	ADD PLACEMENT POLICY CONSTRAINTS="[+engine=tiflash]" ROLE=learner REPLICAS=1;
+CREATE TABLE t1 (
+	id INT NOT NULL PRIMARY KEY,
+	b VARCHAR(100)
+);
+ALTER TABLE t1 PLACEMENT CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=1, CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1, CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1;
 ```
 
-The only way to judge whether it’s adding a TiFlash replica is to check the label. If it contains `engine=tiflash`, then it’s adding or removing a TiFlash replica. This logic is conventional in PD for now.
+#### Placement Policy
 
-Placement rules must conform to Raft constraints. For example, an error should be reported when executing this statement:
+Creating a new `PLACEMENT POLICY`:
 
 ```sql
-ALTER TABLE test
-	ALTER PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=2;
+CREATE PLACEMENT POLICY `companystandardpolicy` CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=1, CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1, CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1;
 ```
 
-There can only be one leader, so `REPLICAS` must be 1 or omitted. But for other roles, `REPLICAS` must be specified.
-
-Besides, at most one role can be defined on the same object. If multiple rules are added on the same role, they will be combined to one rule. For example:
+Creating a new table with a `PLACEMENT POLICY`:
 
 ```sql
-ALTER TABLE test
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=voter REPLICAS=2,
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=bj]" ROLE=voter REPLICAS=2;
+CREATE TABLE t1 (
+	id INT NOT NULL PRIMARY KEY,
+	b VARCHAR(100)
+) PLACEMENT POLICY=`companystandardpolicy`;
 ```
 
-The same role `voter` is defined in 2 different rules, each of which adds 2 replicas. So it is equivalent to:
+Adding `PLACEMENT POLICY` to an existing table:
 
 ```sql
-ALTER TABLE test
-	ADD PLACEMENT POLICY CONSTRAINTS="{+zone=sh:2,+zone=bj:2}" ROLE=voter REPLICAS=4;
+CREATE TABLE t1 (
+	id INT NOT NULL PRIMARY KEY,
+	b VARCHAR(100)
+);
+ALTER TABLE t1 PLACEMENT POLICY=`companystandardpolicy`;
```

Comment:
I'm always wondering when should there be a `SET` and `=`.
E.g. setting tiflash replicas: `ALTER TABLE table_name SET TIFLASH REPLICA count`.
I think it should also be consistent with setting region labels in the future.

Reply:
In this instance, I copied the `PLACEMENT` (removing `ALTER PLACEMENT`) from [`CREATE TABLE`](https://dev.mysql.com/doc/refman/8.0/en/create-table.html) and [`ALTER TABLE`](https://dev.mysql.com/doc/refman/8.0/en/alter-table.html) considered it a _table_option_.

For `POLICY=x` versus POLICY 'x'`, the table options are actually inconsistent with this (most do use `=`, but `TABLESPACE` for example does not). It also made sense to use `POLICY=` so this explicitly refered to a named global `POLICY` versus `CONSTRAINTS=` refers to an inline policy.

For `SET TIFLASH REPLICA count`, I agree this is actually inconsistent with MySQL. The only scenario in an `ALTER TABLE` where `SET` is used, is in `CHARACTER SET`, or in `ALTER TABLE.. ALTER COLUMN`.

I plan to review the region labels feature tomorrow and make this same comment.